kernels/group_gemm/group_gemm.cu

```diff
@@ -0,0 +1,263 @@
+    static constexpr int M_BLOCK = _M_BLOCK, SUPER_M = _SUPER_M, CLUSTER_BLOCKS = 2;
+    using layout    = matmul_layout<M_BLOCK>;
+    static constexpr int NUM_CONSUMER_WARPS=M_BLOCK*4, INPUT_PIPE_STAGES=4, PRODUCER_BARRIER_ARRIVALS=1;
+    // Helper functions
+    template<bool PERISISTENT_GRID=true> __host__ static inline dim3 grid(int M, int N, int K) {
+        return dim3(PERISISTENT_GRID ? 128 : M*N/(M_BLOCK*layout::c_tile::num_elements));
+    }
+    // ThunderKittens template functions
+    __device__ static inline void common_setup(common_setup_args<layout> args, bool is_prepared = false) {
+        if (is_prepared) {
+            return;
+        }
+
+        auto my_scheduler = deep_gemm::Scheduler<deep_gemm::GemmType::GroupedContiguous, static_cast<uint32_t>(M_BLOCK*layout::c_tile::rows),
+                    static_cast<uint32_t>(layout::c_tile::cols), 2, false>(
+                        static_cast<uint32_t>(args.globals.C.rows()),
+                        static_cast<uint32_t>(args.globals.C.cols()),
+                        static_cast<uint32_t>(args.globals.B.depth()),
+                        args.globals.index.raw_ptr);
+        bool is_valid = my_scheduler.get_next_block(
+            args.common.block_m_idx, args.common.block_n_idx, args.task_iter
+        );
+        if (is_valid) {
+            args.num_iters = args.globals.A.cols() / layout::a_tile::cols;
+            int id = warpgroup::groupid() == NUM_CONSUMER_WARPS/4 ? 0 : warpgroup::groupid();
+            args.common.is_tma_multicast_valid = my_scheduler.is_tma_multicast_valid(args.common.block_m_idx);
+            args.common.block_m_idx = args.common.block_m_idx * M_BLOCK + id;
+            args.common.group_idx = args.globals.index.raw_ptr[args.common.block_m_idx  * layout::c_tile::rows];
+        } else {
+            args.num_iters = -1; // No more work to do
+            return;
+        }
+    }
+
+    struct producer {
+        __device__ static void setup(producer_setup_args<layout> args) {
+            warpgroup::decrease_registers<40>(); // decrease registers for producers
+        }
+        __device__ static void load(producer_load_cluster_args<layout> args) {
+            if(warpgroup::warpid() == 0) {
+                tma::expect(args.inputs_arrived, args.input.a, args.input.scale_a_sv);
+                #pragma unroll
+                for(int i = 0; i < M_BLOCK; i++) {
+                    tma::load_async(args.input.a[i], args.globals.A,
+                                    {args.common.block_m_idx+i, args.iter}, args.inputs_arrived);
+                    tma::load_async(args.input.scale_a_sv[i], args.globals.scale_a, {args.iter, args.common.block_m_idx+i}, args.inputs_arrived);
+                }
+
+                if (args.common.is_tma_multicast_valid) {
+                    if (cluster_ctarank() == 0) {
+                        tma::cluster::expect(args.inputs_cluster_arrived, 0, args.input.b);
```

Comment:
@benjaminfspector @simran-arora Currently, the implementation only reaches 80% of the performance of deepgeem. I am confused that the TMA multicast doesn't get a performance improvement, as deepgemm TMA multicast can achieve a 20% performance improvement. Could you help see whether the usage of TMA cluster load is okay or not? Thanks!

Reply:
@DanFu09 @benjaminfspector @simran-arora Any suggestions? Thanks!